core/trino-main/src/main/java/io/trino/sql/planner/LocalExecutionPlanner.java

```diff
@@ -3497,24 +3503,29 @@ private int getWriterCount(Session session, WriterScalingOptions connectorScalin
                 return 1;
             }
 
+            int maxWritersBasedOnMemory = getMaxPartitionWritersBasedOnMemory(session);
             if (partitioningScheme.isPresent()) {
                 // The default value of partitioned writer count is 32 which is high enough to use it
                 // for both cases when scaling is enabled or not. Additionally, it doesn't lead to too many
                 // small files since when scaling is disabled only single writer will handle a single partition.
+                int partitionedWriterCount = getTaskWriterCount(session);
                 if (isLocalScaledWriterExchange(source)) {
-                    return connectorScalingOptions.perTaskMaxScaledWriterCount()
+                    partitionedWriterCount = connectorScalingOptions.perTaskMaxScaledWriterCount()
                             .map(writerCount -> min(writerCount, getTaskPartitionedWriterCount(session)))
                             .orElse(getTaskPartitionedWriterCount(session));
                 }
-                return getTaskPartitionedWriterCount(session);
+                // Consider memory while calculating writer count.
```

Comment:
you could extract `min(partitionedWriterCount, previousPowerOfTwo(maxWritersBasedOnMemory))` and use it here and in `plan` method. Then it would be obvious it's same thing and where it comes from